---
title: "Ovarian and Breast Cancer Pleiotropy analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
require(dmetar)
library(MRcML)
library("IMRP")
library(tidyverse)
library(dplyr)
library(MRMix)
library(MRPRESSO)
```

## Ovarian and Breast Cancer

**Data source**

- Exposure: 17 SNPs from 2015 GWAS on invasive epithelial ovarian cancer, [Identification of six new susceptibility loci for invasive epithelial ovarian cancer](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4445140/).
- Outcome: Breast Cancer Association Consortium 2020 GWAS on Breast Cancer [Zhang H et al (Nat Genet, 2020)](https://www.ncbi.nlm.nih.gov/pubmed/32424353). 


> Step 1. Load and tidy data

```{r}
Ovarian_BC<-read.csv("Data/Ovarian_BC.csv")
Ovarian_BC<-Ovarian_BC[Ovarian_BC$Index.SNP!="rs4691139",]
Ovarian_BC$N<-46282
Ovarian<-se.from.p(effect.size = Ovarian_BC$OR, p = Ovarian_BC$P, N =Ovarian_BC$N,
                      effect.size.type = "ratio")
```

There are 16 snps in both exposure and outcome dataset. We remove rs4691139 because it is not significant with BC in the discovery dataset.

> Step 2. Merge data

Merge datasets for the exposure and the outcome, compute beta.y from odds ratio (OR). Keep the following variables: SNP ID (SNP), chromosome (chr), base pair (bp), effect and non-effect alleles (EA.x, NEA.x, EA,y, NEA.y), sample size of the study associated with X (nx), estimate of genetic effect (beta.x, beta.y) and standard errors (se.x, se.y), effect allele frequency (EAF.x, EAF.y). Remove the SNPs of which the alleles do not match between the exposure and outcome datasets.

```{r}
data = Ovarian_BC %>% bind_cols(Ovarian) %>%
  mutate(beta.y = Beta.meta) %>%
  select(SNP= Index.SNP, chr = Chromosome, bp=Positionb, EA.x = Eff, beta.x = logEffectSize, 
         se.x = logStandardError, EAF.x = EAF,nx=N,
         EA.y = Effect.Meta, NEA.y = Baseline.Meta, beta.y, se.y = sdE.meta, EAF.y = Freq.Gwas) 
# check allele 
data %>% filter(!((EA.x==EA.y) | (EA.x==NEA.y))) %>% nrow()
```

There are 4 snps which the alleles do not match between the exposure and outcome datasets, they may due to strand flips. keep them in the dataset for now.

>Step 3. Harmonize data

Check the allele frequency of palindromic SNPs. Results show that the are coded with respect to the same strand.

```{r}
data %>% filter((EA.y=="A"&NEA.y=="T") | (EA.y=="T"&NEA.y=="A") | (EA.y=="G"&NEA.y=="C") | (EA.y=="C"&NEA.y=="G")) %>%
    select(SNP, EA.x, EAF.x, EA.y, NEA.y, EAF.y)
```




Flip the sign of beta.y if the effect allele in the study associated with X is the non-effect allele in the study associated with Y.
*Note: We did not flip any sign because the allele frequency matched, further discussion needed.* 

>Step 4. Standardize data

Since the exposure (Ovarian) and outcome (BC) are all binary, the summary statistics are standardized by genotypic variance calculated as $2*MAF*(1-MAF)$ under Hardy-Weinberg equilibrium.

```{r}
data_std = with(data, standardize(beta.x,beta.y,se.x,se.y,xtype = "binary", 
                                  ytype = "binary", nx = NULL, ny = NULL, MAF = EAF.y))
```   

> Step 5. Analysis

1. cML-MA Method

```{r}
cML_result = mr_cML(data$beta.x,
                    data$beta.y,
                    data$se.x,
                    data$se.y,
                    n = 46282,
                    random_start = 100,
                    random_seed = 1)
cML_result
```

2. MRMix analysis

```{r}
res = MRMix(data_std$betahat_x_std, data_std$betahat_y_std, 
            data_std$sx_std, data_std$sy_std, profile = TRUE)
str(res)
```

3. MR-Egger

```{r}
Egger<-MREgger(BetaYG="beta.y", BetaXG="beta.x",seBetaYG="se.y", data=data)
Egger
```

4. MR-Presso
```{r}
presso<-mr_presso(BetaOutcome = "beta.y", BetaExposure = "beta.x", SdOutcome = "se.y", SdExposure = "se.x", OUTLIERtest = TRUE, DISTORTIONtest = TRUE, data = data, NbDistribution = 1000,  SignifThreshold = 0.05)
presso
```

